debian/wlk2sqlite.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH WLK2SQLITE 1 "November 19, 2009"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
wlk2sqlite \- program to convert Weatherlink WLK archive files to wview SQLite3 archive records 
.SH SYNOPSIS
.B wlk2sqlite
.RI source_directory [dest_directory]
.br
.SH DESCRIPTION
This manual page describes the use of
.B wlk2sqlite
.
Convert WLK archive file data in source_directory to SQLite3 format in 
[destination_directory]/wview-archive.sdb or to $localstatedir/var/wview/archive/wview-archive.sdb 
if [destination_directory] is not given. Only unique records will be inserted.
.BR Note: Must be run as root if [destination_directory] is not given.
.BR Note: source_directory can be the same as [destination_directory].
.SH OPTIONS
There are no options and source_directory is mandatory.
.TP
.SH SEE ALSO
.BR sqlite2wlk (1),
.BR /usr/share/doc/wview/wview-User-Manual.html
.br
.SH AUTHOR
wlk2sqlite was written by Mark Teel.
.PP
This manual page was written by Mark Teel <mteel2005@gmail.com>,
for the Debian project (and may be used by others).


ChangeLog
RELEASE VERSION HISTORY (newest to oldest)
------------------------------------------


5.21.5    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Remove debug message.




5.21.4    04-06-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Added stale pid file cleanup to the APT package start script.

2)  WMRUSB HID interface changes. Changed reader thread to blocking IO.
    Mitigated the console occasionally blasting out bogus data. Now zero
    packet loss. Fixed issue with pressure sanity check. Added sanity
    check for bogus rain accumulator values from the WMR console.




5.21.3    03-31-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Further WMRUSB interface improvements. Better packet framing.




5.21.2    03-30-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Updated debian APT postinst script to include copying the exfoliation skin
    for new installs.

2)  Enabled checksum processing for WMR stations.

3)  Changed LOOP windspeed items to floats for better precision.
    NOTE: If you use custom datafeed clients the LOOP data structure has changed.

4)  Added temp only packet support for WMRUSB stations. This adds support for
    add-on sensors such as the pool sensor.

5)  Fixed APT installs changing the wview.log file ownership incorrectly. Also
    restarts rsyslog automatically if a new install.




5.21.1    03-13-2014
--------------------

radlib version required: 2.12.0 or newer

1)  Modified HILOW database storage so higher precision is used for cumulative
    data such as rain and ET.

2)  Fixed issue with yearly historical charts when generated at wview start
    (as opposed to at midnight). The current day was erroneously being included in
    the history data. This was most apparent with the UV and Solar Radiation
    charts.

3)  Fixed bug introduced in 5.21.0 release causing NULL wind direction to have
    360 added to it (to normalize calibration result). A value of -99640 was
    being stored/displayed.




5.21.0    03-05-2014
--------------------

radlib version required: 2.12.0 or newer

1)  WMRUSB improvements. Further reverse engineering of the undocumented protocol. Reset and
    heartbeat messages now working. Packet reception more stable. Initial packet acquisition
    faster resulting in faster startup.

2)  Modify CWOP packet header for new stricter compliance requirements of CWOP servers.

3)  Fix log message time_t bug for 64-bit platforms.

4)  Add timeout parameter for ssh rules. Added rows to config database.

5)  Updated data feed conversion logic to handle ARCHIVE_VALUE_NULL in loop data.

6)  Added flat line sensor value detection for outside temp, wind direction and wind speed.
    If there are no changes over a 4 hour period an email alert will be sent.

7)  Added Station Name, City and State to the wviewmgmt System Status page header.

8)  Added logic to SSH timer to fixup next timer interval if system time has skewed or
    if the ssh transfer took longer than 2 minutes.

9)  Fixed TE923 rain calculation to assume each tip is 1/36" instead of 1mm.

10) Fixed a bug for Vantage Pro stations which could cause the last archive interval of
    the previous day to be included in the current day HILOW data. This occurred if the
    midnight archive record was provided late by the console.

11) Added wet bulb temperature calculation and html tag --WetBulbTemp--. Added to all
    "readings.incx" templates.

12) Added email alert notification if no loop data has been received for a given archive
    interval. This applies to all stations except the Davis stations (alerts already
    exist for them).

13) Updated start scripts to detect and delete stale pid files.

14) Modified WH1080 data parsing to assume a humidity of 9% if the value from the
    station is undefined (0xff).




5.20.2    03-21-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added back "dailyrainin" for http reporting. They do no computation based on hourly rain values.

2)  Fixed a typo in the wviewmgmt station.php file. 




5.20.1    03-20-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Modified http rain submission (Wunderground and Weatherforyou) so only the
    "rainin" parameter is sent so daily rain will be correct. 




5.20.0    03-17-2012
--------------------

radlib version required: 2.12.0 or newer

1)  Added "wait_for_time_set" function to most wview start script examples so 
    wview daemons are not started until the system time is set. This applies to
    platforms without a real time clock which must reset their time during 
    startup.

2)  Updated the User Manual section on skin selection to better describe enabling
    extended data generation (UV, ET, radiation).

3)  Updated the debian post install script to try harder to find the default http
    document root location for creation of symbolic links to wviewmgmt and the
    weather home page.

4)  Added new configuration option to optionally disable using the Vantage Pro 
    console archive records and instead autogenerate them from the LOOP data.
    This is how all other stations work. It also supports non-standard console
    interfaces (those not using the datalogger on the console).

5)  Updated the Weatherforyou URL for data submissions.

6)  Fixed bug with DataFeed clients that caused an infinite loop if you
    restarted wview while there was a datafeed client connection.

7)  Fixed bug with Alarm Type configuration introduced in 5.19.0.

8)  Changed TWI station to compute rain fall rate as it is reported to not
    be provided by the station.

9)  Added socket retry logic to the station ethernet interface (used by the
    Vantage Pro IP datalogger and any serial-to-ethernet device).

10) Modified CWOP host name to IP address resolution so that 'getaddrinfo' is
    used instead of 'gethostbyname' and all results are tried in the order they 

utilities/wlk2sqlite/wlk2sqlite.c
/*---------------------------------------------------------------------
 
 FILE NAME:
        wlk2sqlite.c
 
 PURPOSE:
        wview archive file convertor utility: Davis WLK TO SQLite3 database.
 
 REVISION HISTORY:
    Date        Programmer  Revision    Function
    09/13/2008  M.S. Teel   0           Original
 
 ASSUMPTIONS:
 None.
 
------------------------------------------------------------------------*/
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/time.h>
#include <sys/select.h>

#include <sysdefs.h>
#include <datadefs.h>
#include <dbfiles.h>
#include "wvutilities.h"


static void USAGE (void)
{
    printf ("Usage: wlk2sqlite <source_directory> [destination_directory]\n\n");
    printf ("    Convert WLK archive file data in <source_directory> to SQLite3 format in [destination_directory]/wview-archive.sdb\n");
    printf ("    or to $prefix/var/wview/archive/wview-archive.sdb if [destination_directory] is not given\n");
    printf ("Note: Must be run as root if [destination_directory] is not given. Only unique records will be inserted.\n");
    printf ("Note: <source_directory> can be the same as [destination_directory]\n\n");
    return;
}

static void convertWLKToArchivePkt(ArchiveRecord* newRecord, ARCHIVE_PKT* archivePkt, time_t timeStamp)
{
    Data_Indices index;
    float           click;

    // create the time_t time for the record:
    archivePkt->dateTime = (int32_t)timeStamp;

    archivePkt->usUnits  = 1;
    archivePkt->interval = newRecord->archiveInterval;

    // Set all values to NULL by default:
    for (index = DATA_INDEX_barometer; index < DATA_INDEX_MAX; index ++)
    {
        archivePkt->value[index] = ARCHIVE_VALUE_NULL;
    }

    // Set the values we can:
    archivePkt->value[DATA_INDEX_outTemp]        = (float)newRecord->outsideTemp/10.0;
    archivePkt->value[DATA_INDEX_barometer]      = (float)newRecord->barometer/1000.0;
    archivePkt->value[DATA_INDEX_inTemp]         = (float)newRecord->insideTemp/10.0;
    archivePkt->value[DATA_INDEX_inHumidity]     = (float)newRecord->insideHum/10.0;
    archivePkt->value[DATA_INDEX_outHumidity]    = (float)newRecord->outsideHum/10.0;
    archivePkt->value[DATA_INDEX_windSpeed]      = (float)newRecord->windSpeed/10.0;
    if (newRecord->windDirection < 16)
    {
        archivePkt->value[DATA_INDEX_windDir]    = (float)newRecord->windDirection * 22.5;
    }
    archivePkt->value[DATA_INDEX_windGust]       = (float)newRecord->hiWindSpeed/10.0;
    if (newRecord->hiWindDirection < 16)
    {
        archivePkt->value[DATA_INDEX_windGustDir] = (float)newRecord->hiWindDirection * 22.5;
    }

    if ((newRecord->rain & 0xF000) == 0x0000)
        click = 10;
    else if ((newRecord->rain & 0xF000) == 0x2000) //0.2 mm 0x2000
        click = 127;
    else if ((newRecord->rain & 0xF000) == 0x3000) //1.0 mm 0x3000
        click = 25.4;
    else if ((newRecord->rain & 0xF000) == 0x6000) //0.1 mm 0x6000 (not fully supported)
        click = 254;
    else
        click = 100;

    archivePkt->value[DATA_INDEX_rainRate]       
        = (float)newRecord->hiRainRate/click;
    archivePkt->value[DATA_INDEX_rain]           
        = ((float)(newRecord->rain & 0xFFF))/click;
    archivePkt->value[DATA_INDEX_dewpoint]       
        = wvutilsCalculateDewpoint ((float)archivePkt->value[DATA_INDEX_outTemp],
                                    (float)archivePkt->value[DATA_INDEX_outHumidity]);
    archivePkt->value[DATA_INDEX_windchill]      
        = wvutilsCalculateWindChill ((float)archivePkt->value[DATA_INDEX_outTemp],
                                     (float)archivePkt->value[DATA_INDEX_windSpeed]);
    archivePkt->value[DATA_INDEX_heatindex]      
        = wvutilsCalculateHeatIndex ((float)archivePkt->value[DATA_INDEX_outTemp],
                                     (float)archivePkt->value[DATA_INDEX_outHumidity]);
    archivePkt->value[DATA_INDEX_ET]             = (float)newRecord->ET/1000.0;
    if ((uint16_t)newRecord->solarRad != 0x7FFF && 
        (uint16_t)newRecord->solarRad != 0xFFFF && 
        (float)newRecord->solarRad >= 0 && 
        (float)newRecord->solarRad <= 1800)
    {
        archivePkt->value[DATA_INDEX_radiation]  = (float)newRecord->solarRad;
    }
    if (newRecord->UV != 0xFF)
    {
        archivePkt->value[DATA_INDEX_UV]         = (float)newRecord->UV/10.0;
    }
    if (newRecord->extraTemp[0] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_extraTemp1] = (float)(newRecord->extraTemp[0] - 90);
    }
    if (newRecord->extraTemp[1] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_extraTemp2] = (float)(newRecord->extraTemp[1] - 90);
    }
    if (newRecord->extraTemp[2] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_extraTemp3] = (float)(newRecord->extraTemp[2] - 90);
    }
    if (newRecord->soilTemp[0] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilTemp1]  = (float)(newRecord->soilTemp[0] - 90);
    }
    if (newRecord->soilTemp[1] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilTemp2]  = (float)(newRecord->soilTemp[1] - 90);
    }
    if (newRecord->soilTemp[2] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilTemp3]  = (float)(newRecord->soilTemp[2] - 90);
    }
    if (newRecord->soilTemp[3] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilTemp4]  = (float)(newRecord->soilTemp[3] - 90);
    }
    if (newRecord->leafTemp[0] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_leafTemp1]  = (float)(newRecord->leafTemp[0] - 90);
    }
    if (newRecord->leafTemp[1] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_leafTemp2]  = (float)(newRecord->leafTemp[1] - 90);
    }
    if (newRecord->extraHum[0] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_extraHumid1] = (float)newRecord->extraHum[0];
    }
    if (newRecord->extraHum[1] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_extraHumid2] = (float)newRecord->extraHum[1];
    }
    if (newRecord->soilMoisture[0] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilMoist1] = (float)newRecord->soilMoisture[0];
    }
    if (newRecord->soilMoisture[1] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilMoist2] = (float)newRecord->soilMoisture[1];
    }
    if (newRecord->soilMoisture[2] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilMoist3] = (float)newRecord->soilMoisture[2];
    }
    if (newRecord->soilMoisture[3] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_soilMoist4] = (float)newRecord->soilMoisture[3];
    }
    if (newRecord->leafWetness[0] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_leafWet1]   = (float)newRecord->leafWetness[0];
    }
    if (newRecord->leafWetness[1] != 0xFF)
    {
        archivePkt->value[DATA_INDEX_leafWet2]   = (float)newRecord->leafWetness[1];
    }

    return;
}

static void ConvertWlkToSqlite (char *srcDir, char *destDir)
{
    time_t              timeStamp, diffTime, startTime = time(NULL);
    struct tm           buildTime;
    uint16_t            recdate, rectime;
    ArchiveRecord       wlkPkt;
    ARCHIVE_PKT         sqlitePkt;
    int                 dups = 0, inserts = 0, errors = 0, lastminute = 0;

    if (destDir != NULL)
    {
        dbsqliteArchiveSetPath(destDir);
    }

    // ... Initialize the archive database interface:
    if (dbsqliteArchiveInit() == ERROR)
    {
        printf("dbsqliteArchiveInit failed\n");
        return;
    }
    
    // Avoid expensive fsyncs during conversion
    if (dbsqliteArchivePragmaSet("synchronous", "off") == ERROR)
    {
        printf("dbsqliteArchivePragmaSet synchronous = off failed\n");
        return;
    }

    // Avoid expensive journal operations during conversion
    if (dbsqliteArchivePragmaSet("journal_mode", "off") == ERROR)
    {
        printf("dbsqliteArchivePragmaSet journal_mode = off failed\n");
        return;
    }

    // Set the start date to 1/1/2000:
    recdate = INSERT_PACKED_DATE(2000,1,1);
    rectime = 0;
    while (dbfGetNextArchiveRecord(srcDir, &recdate, &rectime, &wlkPkt) != ERROR)
    {
        // Build the sqlite timestamp:
        memset (&buildTime, 0, sizeof(buildTime));
        buildTime.tm_min    = EXTRACT_PACKED_MINUTE(rectime);
        buildTime.tm_hour   = EXTRACT_PACKED_HOUR(rectime);
        buildTime.tm_mday   = EXTRACT_PACKED_DAY(recdate);
        buildTime.tm_mon    = EXTRACT_PACKED_MONTH(recdate) - 1;
        buildTime.tm_year   = EXTRACT_PACKED_YEAR(recdate) - 1900;
        buildTime.tm_isdst  = -1;
        timeStamp = mktime(&buildTime);

        // Check to see if this record exists in the sqlite DB:
        if (dbsqliteArchiveGetRecord(timeStamp, &sqlitePkt) != ERROR)
        {
            // duplicate, skip it:
            dups ++;
            continue;
        }

        // OK, need to insert this puppy:
        convertWLKToArchivePkt(&wlkPkt, &sqlitePkt, timeStamp);

        if (dbsqliteArchiveStoreRecord(&sqlitePkt) == ERROR)
        {
            // Error:
            errors ++;
        }
        else
        {
            // Good stuff:
            inserts ++;
        }

        diffTime = time(NULL) - startTime;

        // Output running stats
        if ( (int)diffTime%60 == 0 && (int)diffTime/60 > 0 && (int)diffTime/60 != lastminute )
        {
            printf("Conversion Stats:\n");
            printf("    Time      : %d:%d minutes\n", (int)diffTime/60, (int)diffTime%60);
            printf("    Recs/sec  : %d\n", (inserts+dups+errors)/(int)diffTime);
            printf("    Inserts   : %d\n", inserts);
            printf("    Duplicates: %d\n", dups);
            printf("    Errors    : %d\n", errors);
            lastminute = (int)diffTime/60;
        }
    }

    dbsqliteArchiveExit();

    diffTime = time(NULL) - startTime;

    // Output results:
    printf("Conversion Stats:\n");
    printf("    Time      : %d:%d\n", (int)diffTime/60, (int)diffTime%60);
    printf("    Recs/sec  : %d\n", (inserts+dups+errors)/(int)diffTime);
    printf("    Inserts   : %d\n", inserts);
    printf("    Duplicates: %d\n", dups);
    printf("    Errors    : %d\n", errors);
}

////////////////////////////////////////////////////////////////////////////////
int main (int argc, char *argv[])
{
    char            *SourceDir, *DestDir;
    struct stat     fileData;
    
    if (argc < 2)
    {
        USAGE ();
        exit (-1);
    }
    if (argc < 3)
    {
        printf("Using default $prefix/var/wview/archive destination...\n");
        DestDir = NULL;
    }
    else
    {
        DestDir = argv[2];
    }

    SourceDir = argv[1];

    // sanity check the arguments
    if (stat(SourceDir, &fileData) != 0)
    {
        printf ("Source directory %s does not exist!\n", SourceDir);
        exit (-1);
    }
    else if (!(fileData.st_mode & S_IFDIR))
    {
        printf ("Source directory %s is not a directory!\n", SourceDir);
        exit (-1);
    }
    if (DestDir)
    {
        if (stat(DestDir, &fileData) != 0)
        {
            printf ("Destination directory %s does not exist!\n", DestDir);
            exit (-1);
        }
        else if (!(fileData.st_mode & S_IFDIR))
        {
            printf ("Destination directory %s is not a directory!\n", DestDir);
            exit (-1);
        }
    }

    // OK, args appear to be good, convert WLK files to our SQLite DB:
    printf("Converting...(this takes a while, longer on small devices)...\n");
    ConvertWlkToSqlite (SourceDir, DestDir);
    exit (0);
}

